From c79383c49f19ff0721e9572b36916d435f550a31 Mon Sep 17 00:00:00 2001
Message-ID: <c79383c49f19ff0721e9572b36916d435f550a31.1779975966.git.bertrand.marquis@arm.com>
In-Reply-To: <55d56a174a0f04520d4a1558630bb0b739074b13.1779975966.git.bertrand.marquis@arm.com>
References: <55d56a174a0f04520d4a1558630bb0b739074b13.1779975966.git.bertrand.marquis@arm.com>
From: Bertrand Marquis <bertrand.marquis@arm.com>
Date: Fri, 22 May 2026 11:48:44 +0200
Subject: [PATCH 15/30] virtio-msg: Sync transport revision 1 message layouts

Update the local virtio-msg wire structures and pack/unpack helpers to
match transport revision 1. This includes the device-info UUID and feature
block fields, feature block arrays, vqueue flags, shared-memory region
fields, bus response tokens, bus event naming, and generated message type
masking.

Keep runtime policy changes in later patches so reviewers can separate wire
compatibility from behavior. QEMU still uses a local 128-byte storage
envelope here; transports advertise the active maximum message size
separately through their latched capabilities.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>
---
 hw/virtio/virtio-msg.c              |  35 +++--
 include/hw/virtio/virtio-msg-prot.h | 236 +++++++++++++++++++---------
 2 files changed, 181 insertions(+), 90 deletions(-)

diff --git a/hw/virtio/virtio-msg.c b/hw/virtio/virtio-msg.c
index 7b0a68cef133..bd7a333bc791 100644
--- a/hw/virtio/virtio-msg.c
+++ b/hw/virtio/virtio-msg.c
@@ -50,7 +50,7 @@ static bool virtio_msg_bad(VirtIOMSGProxy *s, VirtIOMSG *msg)
         min_size += sizeof msg->get_features;
         break;
     case VIRTIO_MSG_SET_FEATURES:
-        n = msg->set_features.num;
+        n = msg->set_features.num_blocks;
 
         /* We expect at least one feature block.  */
         if (n == 0 || n > VIRTIO_MSG_MAX_FEATURE_NUM) {
@@ -148,7 +148,8 @@ static void virtio_msg_bus_get_devices(VirtIOMSGProxy *s,
     }
 
     virtio_msg_pack_bus_get_devices_resp(&msg_resp,
-                                         offset, num, next_offset,
+                                         offset, next_offset, num,
+                                         msg->token,
                                          data);
     virtio_msg_send_must_deliver(s, &msg_resp);
 }
@@ -157,7 +158,7 @@ static void virtio_msg_bus_ping(VirtIOMSGProxy *s, VirtIOMSG *msg)
 {
     VirtIOMSG msg_resp;
 
-    virtio_msg_pack_bus_ping_resp(&msg_resp, msg->bus_ping.data);
+    virtio_msg_pack_bus_ping_resp(&msg_resp, msg->token, msg->bus_ping.data);
     virtio_msg_send_must_deliver(s, &msg_resp);
 }
 
@@ -167,13 +168,14 @@ static void virtio_msg_device_info(VirtIOMSGProxy *s,
     VirtIODevice *vdev = virtio_msg_vdev(s, msg->dev_num);
     uint32_t config_len = 0;
     uint32_t device_id = 0;
-    uint32_t max_vqs = 0;
+    uint32_t max_virtqueues = 0;
+    uint8_t nil_uuid[16] = { 0 };
     VirtIOMSG msg_resp;
 
     if (vdev) {
         device_id = vdev->device_id;
         config_len = vdev->config_len;
-        max_vqs = virtio_get_num_queues(vdev);
+        max_virtqueues = virtio_get_num_queues(vdev);
     } else {
         error_report("%s: No virtio device on bus %s!",
                      __func__, BUS(&s->devs[msg->dev_num].bus)->name);
@@ -182,10 +184,10 @@ static void virtio_msg_device_info(VirtIOMSGProxy *s,
     virtio_msg_pack_get_device_info_resp(&msg_resp, msg->dev_num, msg->token,
                                          device_id,
                                          VIRTIO_MSG_VENDOR_ID,
-                                         /* Feature bits */
-                                         64,
+                                         nil_uuid,
+                                         2,
                                          config_len,
-                                         max_vqs,
+                                         max_virtqueues,
                                          0, 0);
     virtio_msg_send_must_deliver(s, &msg_resp);
 }
@@ -195,9 +197,10 @@ static void virtio_msg_get_features(VirtIOMSGProxy *s,
 {
     VirtIODevice *vdev = virtio_msg_lookup_vdev(s, msg->dev_num, __func__);
     VirtIOMSG msg_resp;
-    uint32_t index = msg->get_features.index;
+    uint32_t index = msg->get_features.block_index;
     uint32_t f[VIRTIO_MSG_MAX_FEATURE_NUM] = { 0 };
-    uint32_t num = MIN(msg->get_features.num, VIRTIO_MSG_MAX_FEATURE_NUM);
+    uint32_t num = MIN(msg->get_features.num_blocks,
+                       VIRTIO_MSG_MAX_FEATURE_NUM);
     uint64_t features = 0;
 
     if (vdev) {
@@ -228,15 +231,16 @@ static void virtio_msg_set_features(VirtIOMSGProxy *s,
 
     f = s->devs[msg->dev_num].guest_features;
 
-    for (i = 0; i < msg->set_features.num; i++) {
-        unsigned int feature_index = i + msg->set_features.index;
+    for (i = 0; i < msg->set_features.num_blocks; i++) {
+        unsigned int feature_index = i + msg->set_features.block_index;
 
         /* We only support up to 64bits  */
         if (feature_index >= 2) {
             break;
         }
 
-        f = deposit64(f, feature_index * 32, 32, msg->set_features.b32[i]);
+        f = deposit64(f, feature_index * 32, 32,
+                      msg->set_features.features[i]);
     }
 
     s->devs[msg->dev_num].guest_features = f;
@@ -371,12 +375,12 @@ static void virtio_msg_get_vqueue(VirtIOMSGProxy *s,
             virtio_queue_get_rings(vdev, index, &desc, &avail, &used);
         }
         virtio_msg_pack_get_vqueue_resp(&msg_resp, msg->dev_num, msg->token,
-                                        index, max_size, size,
+                                        index, max_size, size, 0,
                                         desc, avail, used);
     } else {
         /* OOB index or missing device, respond with all zeroes. */
         virtio_msg_pack_get_vqueue_resp(&msg_resp, msg->dev_num, msg->token,
-                                        index, 0, 0, 0, 0, 0);
+                                        index, 0, 0, 0, 0, 0, 0);
     }
 
     virtio_msg_send_must_deliver(s, &msg_resp);
@@ -987,6 +991,7 @@ static void virtio_msg_realize(DeviceState *d, Error **errp)
     VirtIOMSGProxy *s = VIRTIO_MSG(d);
     Object *o = OBJECT(d);
     int i;
+
     virtio_msg_reset_transport_caps(s);
 
     for (i = 0; i < ARRAY_SIZE(s->devs); i++) {
diff --git a/include/hw/virtio/virtio-msg-prot.h b/include/hw/virtio/virtio-msg-prot.h
index 89d1d17b9cd5..ce397968e939 100644
--- a/include/hw/virtio/virtio-msg-prot.h
+++ b/include/hw/virtio/virtio-msg-prot.h
@@ -33,7 +33,7 @@ enum {
     VIRTIO_MSG_GET_VQUEUE        = 0x09,
     VIRTIO_MSG_SET_VQUEUE        = 0x0a,
     VIRTIO_MSG_RESET_VQUEUE      = 0x0b,
-    VIRTIO_MSG_GET_SHM           = 0x0c, /* Not yet supported */
+    VIRTIO_MSG_GET_SHM           = 0x0c,
     VIRTIO_MSG_EVENT_CONFIG      = 0x40,
     VIRTIO_MSG_EVENT_AVAIL       = 0x41,
     VIRTIO_MSG_EVENT_USED        = 0x42,
@@ -47,10 +47,33 @@ enum {
 };
 
 #define VIRTIO_MSG_TRANSPORT_REVISION 1
-#define VIRTIO_MSG_MAX_SIZE 48
+#define VIRTIO_MSG_MIN_SIZE 52
+#define VIRTIO_MSG_MAX_SIZE 128
+#define VIRTIO_MSG_MAX_PAYLOAD_SIZE \
+    (VIRTIO_MSG_MAX_SIZE - (sizeof(uint8_t) * 2 + sizeof(uint16_t) * 3))
 
 #define VIRTIO_MSG_TYPE_RESPONSE  (1 << 0)
 #define VIRTIO_MSG_TYPE_BUS       (1 << 1)
+#define VIRTIO_MSG_TYPE_MASK      (VIRTIO_MSG_TYPE_RESPONSE | \
+                                   VIRTIO_MSG_TYPE_BUS)
+
+#define VIRTIO_MSG_VQUEUE_STATE_MASK             0x3U
+#define VIRTIO_MSG_VQUEUE_STATE_KEEP_DISABLED    0x0U
+#define VIRTIO_MSG_VQUEUE_STATE_ENABLE           0x1U
+#define VIRTIO_MSG_VQUEUE_STATE_KEEP             0x2U
+#define VIRTIO_MSG_VQUEUE_STATE_RESERVED         0x3U
+#define VIRTIO_MSG_VQUEUE_F_SIZE_IGNORE          (1U << 2)
+#define VIRTIO_MSG_VQUEUE_F_DESC_ADDR_IGNORE     (1U << 3)
+#define VIRTIO_MSG_VQUEUE_F_DRIVER_ADDR_IGNORE   (1U << 4)
+#define VIRTIO_MSG_VQUEUE_F_DEVICE_ADDR_IGNORE   (1U << 5)
+#define VIRTIO_MSG_VQUEUE_F_IGNORE_MASK \
+    (VIRTIO_MSG_VQUEUE_F_SIZE_IGNORE | \
+     VIRTIO_MSG_VQUEUE_F_DESC_ADDR_IGNORE | \
+     VIRTIO_MSG_VQUEUE_F_DRIVER_ADDR_IGNORE | \
+     VIRTIO_MSG_VQUEUE_F_DEVICE_ADDR_IGNORE)
+#define VIRTIO_MSG_VQUEUE_F_RESERVED_MASK \
+    (~(VIRTIO_MSG_VQUEUE_STATE_MASK | VIRTIO_MSG_VQUEUE_F_IGNORE_MASK))
+#define VIRTIO_MSG_VQUEUE_RESP_F_ENABLED         (1U << 0)
 
 typedef struct VirtIOMSG {
     uint8_t type;
@@ -60,30 +83,31 @@ typedef struct VirtIOMSG {
     uint16_t msg_size;
 
     union {
-        uint8_t payload_u8[40];
+        uint8_t payload_u8[VIRTIO_MSG_MAX_PAYLOAD_SIZE];
 
         struct {
             uint32_t device_id;
             uint32_t vendor_id;
-            uint32_t num_feature_bits;
+            uint8_t device_uuid[16];
+            uint32_t num_feature_blocks;
             uint32_t config_size;
-            uint32_t max_vqs;
-            uint16_t admin_vq_idx;
-            uint16_t admin_vq_count;
+            uint32_t max_virtqueues;
+            uint32_t admin_vq_start;
+            uint32_t admin_vq_count;
         } QEMU_PACKED get_device_info_resp;
         struct {
-            uint32_t index;
-            uint32_t num;
+            uint32_t block_index;
+            uint32_t num_blocks;
         } QEMU_PACKED get_features;
         struct {
-            uint32_t index;
-            uint32_t num;
-            uint32_t b32[];
+            uint32_t block_index;
+            uint32_t num_blocks;
+            uint32_t features[];
         } QEMU_PACKED get_features_resp;
         struct {
-            uint32_t index;
-            uint32_t num;
-            uint32_t b32[];
+            uint32_t block_index;
+            uint32_t num_blocks;
+            uint32_t features[];
         } QEMU_PACKED set_features;
         struct {
             uint32_t offset;
@@ -123,14 +147,14 @@ typedef struct VirtIOMSG {
             uint32_t index;
             uint32_t max_size;
             uint32_t size;
-            uint32_t reserved;
+            uint32_t flags;
             uint64_t descriptor_addr;
             uint64_t driver_addr;
             uint64_t device_addr;
         } QEMU_PACKED get_vqueue_resp;
         struct {
             uint32_t index;
-            uint32_t unused;
+            uint32_t flags;
             uint32_t size;
             uint32_t reserved;
             uint64_t descriptor_addr;
@@ -141,12 +165,13 @@ typedef struct VirtIOMSG {
             uint32_t index;
         } QEMU_PACKED reset_vqueue;
         struct {
-            uint32_t index;
+            uint32_t shmid;
         } QEMU_PACKED get_shm;
         struct {
-            uint32_t index;
-            uint32_t length;
-            uint32_t address;
+            uint32_t shmid;
+            uint32_t reserved;
+            uint64_t length;
+            uint64_t address;
         } QEMU_PACKED get_shm_resp;
         struct {
             uint32_t status;
@@ -170,8 +195,8 @@ typedef struct VirtIOMSG {
         } QEMU_PACKED bus_get_devices;
         struct {
             uint16_t offset;
-            uint16_t num;
             uint16_t next_offset;
+            uint16_t num;
             uint8_t data[];
         } QEMU_PACKED bus_get_devices_resp;
         struct {
@@ -183,7 +208,7 @@ typedef struct VirtIOMSG {
         struct {
             uint16_t dev_num;
 
-#define VIRTIO_MSG_BUS_EVENT_DEVICE_STATE_READY   0x1
+#define VIRTIO_MSG_BUS_EVENT_DEVICE_STATE_ADDED   0x1
 #define VIRTIO_MSG_BUS_EVENT_DEVICE_STATE_REMOVED 0x2
             uint16_t dev_state;
         } QEMU_PACKED bus_event_device;
@@ -192,12 +217,53 @@ typedef struct VirtIOMSG {
 
 /* Maximum number of 32b feature-blocks in a single message.  */
 #define VIRTIO_MSG_MAX_FEATURE_NUM \
-    ((VIRTIO_MSG_MAX_SIZE - offsetof(VirtIOMSG, get_features_resp.b32)) / 4)
+    ((VIRTIO_MSG_MAX_SIZE - offsetof(VirtIOMSG, \
+                                      get_features_resp.features)) / 4)
 
 /* Maximum amount of config-data in a single message, in bytes.  */
 #define VIRTIO_MSG_MAX_CONFIG_BYTES \
     (VIRTIO_MSG_MAX_SIZE - offsetof(VirtIOMSG, set_config.data))
 
+/* Maximum amount of event-config data in a single message, in bytes.  */
+#define VIRTIO_MSG_MAX_EVENT_CONFIG_BYTES \
+    (VIRTIO_MSG_MAX_SIZE - offsetof(VirtIOMSG, event_config.config_value))
+
+static inline unsigned int virtio_msg_max_feature_num_for_size(
+    uint16_t max_msg_size)
+{
+    size_t msg_prefix_size = offsetof(VirtIOMSG, get_features_resp.features);
+
+    if (max_msg_size <= msg_prefix_size) {
+        return 0;
+    }
+
+    return (max_msg_size - msg_prefix_size) / sizeof(uint32_t);
+}
+
+static inline unsigned int virtio_msg_max_config_bytes_for_size(
+    uint16_t max_msg_size)
+{
+    size_t msg_prefix_size = offsetof(VirtIOMSG, set_config.data);
+
+    if (max_msg_size <= msg_prefix_size) {
+        return 0;
+    }
+
+    return max_msg_size - msg_prefix_size;
+}
+
+static inline unsigned int virtio_msg_max_event_config_bytes_for_size(
+    uint16_t max_msg_size)
+{
+    size_t msg_prefix_size = offsetof(VirtIOMSG, event_config.config_value);
+
+    if (max_msg_size <= msg_prefix_size) {
+        return 0;
+    }
+
+    return max_msg_size - msg_prefix_size;
+}
+
 #define LE_TO_CPU(v)                                        \
 {                                                           \
     if (sizeof(v) == 2) {                                   \
@@ -220,9 +286,11 @@ static inline void virtio_msg_unpack_bus(VirtIOMSG *msg)
         break;
     case VIRTIO_MSG_BUS_GET_DEVICES:
         LE_TO_CPU(msg->bus_get_devices.offset);
-        LE_TO_CPU(msg->bus_get_devices.num);
         if (msg->type & VIRTIO_MSG_TYPE_RESPONSE) {
             LE_TO_CPU(msg->bus_get_devices_resp.next_offset);
+            LE_TO_CPU(msg->bus_get_devices_resp.num);
+        } else {
+            LE_TO_CPU(msg->bus_get_devices.num);
         }
         break;
     case VIRTIO_MSG_BUS_EVENT_DEVICE:
@@ -249,18 +317,18 @@ static inline void virtio_msg_unpack_resp(VirtIOMSG *msg)
     case VIRTIO_MSG_DEVICE_INFO:
         LE_TO_CPU(msg->get_device_info_resp.device_id);
         LE_TO_CPU(msg->get_device_info_resp.vendor_id);
-        LE_TO_CPU(msg->get_device_info_resp.num_feature_bits);
+        LE_TO_CPU(msg->get_device_info_resp.num_feature_blocks);
         LE_TO_CPU(msg->get_device_info_resp.config_size);
-        LE_TO_CPU(msg->get_device_info_resp.max_vqs);
-        LE_TO_CPU(msg->get_device_info_resp.admin_vq_idx);
+        LE_TO_CPU(msg->get_device_info_resp.max_virtqueues);
+        LE_TO_CPU(msg->get_device_info_resp.admin_vq_start);
         LE_TO_CPU(msg->get_device_info_resp.admin_vq_count);
         break;
     case VIRTIO_MSG_GET_FEATURES:
-        LE_TO_CPU(msg->get_features_resp.index);
-        LE_TO_CPU(msg->get_features_resp.num);
+        LE_TO_CPU(msg->get_features_resp.block_index);
+        LE_TO_CPU(msg->get_features_resp.num_blocks);
         for (i = 0; i < VIRTIO_MSG_MAX_FEATURE_NUM &&
-                    i < msg->get_features_resp.num; i++) {
-            LE_TO_CPU(msg->get_features_resp.b32[i]);
+                    i < msg->get_features_resp.num_blocks; i++) {
+            LE_TO_CPU(msg->get_features_resp.features[i]);
         }
         break;
     case VIRTIO_MSG_GET_DEVICE_STATUS:
@@ -272,7 +340,8 @@ static inline void virtio_msg_unpack_resp(VirtIOMSG *msg)
         LE_TO_CPU(msg->get_config_resp.size);
         break;
     case VIRTIO_MSG_GET_SHM:
-        LE_TO_CPU(msg->get_shm_resp.index);
+        LE_TO_CPU(msg->get_shm_resp.shmid);
+        LE_TO_CPU(msg->get_shm_resp.reserved);
         LE_TO_CPU(msg->get_shm_resp.length);
         LE_TO_CPU(msg->get_shm_resp.address);
         break;
@@ -285,6 +354,7 @@ static inline void virtio_msg_unpack_resp(VirtIOMSG *msg)
         LE_TO_CPU(msg->get_vqueue_resp.index);
         LE_TO_CPU(msg->get_vqueue_resp.max_size);
         LE_TO_CPU(msg->get_vqueue_resp.size);
+        LE_TO_CPU(msg->get_vqueue_resp.flags);
         LE_TO_CPU(msg->get_vqueue_resp.descriptor_addr);
         LE_TO_CPU(msg->get_vqueue_resp.driver_addr);
         LE_TO_CPU(msg->get_vqueue_resp.device_addr);
@@ -328,15 +398,15 @@ static inline void virtio_msg_unpack(VirtIOMSG *msg)
 
     switch (msg->msg_id) {
     case VIRTIO_MSG_GET_FEATURES:
-        LE_TO_CPU(msg->get_features.index);
-        LE_TO_CPU(msg->get_features.num);
+        LE_TO_CPU(msg->get_features.block_index);
+        LE_TO_CPU(msg->get_features.num_blocks);
         break;
     case VIRTIO_MSG_SET_FEATURES:
-        LE_TO_CPU(msg->set_features.index);
-        LE_TO_CPU(msg->set_features.num);
+        LE_TO_CPU(msg->set_features.block_index);
+        LE_TO_CPU(msg->set_features.num_blocks);
         for (i = 0; i < VIRTIO_MSG_MAX_FEATURE_NUM &&
-                    i < msg->set_features.num; i++) {
-            LE_TO_CPU(msg->set_features.b32[i]);
+                    i < msg->set_features.num_blocks; i++) {
+            LE_TO_CPU(msg->set_features.features[i]);
         }
         break;
     case VIRTIO_MSG_SET_DEVICE_STATUS:
@@ -352,14 +422,16 @@ static inline void virtio_msg_unpack(VirtIOMSG *msg)
         LE_TO_CPU(msg->set_config.size);
         break;
     case VIRTIO_MSG_GET_SHM:
-        LE_TO_CPU(msg->get_shm.index);
+        LE_TO_CPU(msg->get_shm.shmid);
         break;
     case VIRTIO_MSG_GET_VQUEUE:
         LE_TO_CPU(msg->get_vqueue.index);
         break;
     case VIRTIO_MSG_SET_VQUEUE:
         LE_TO_CPU(msg->set_vqueue.index);
+        LE_TO_CPU(msg->set_vqueue.flags);
         LE_TO_CPU(msg->set_vqueue.size);
+        LE_TO_CPU(msg->set_vqueue.reserved);
         LE_TO_CPU(msg->set_vqueue.descriptor_addr);
         LE_TO_CPU(msg->set_vqueue.driver_addr);
         LE_TO_CPU(msg->set_vqueue.device_addr);
@@ -399,7 +471,7 @@ static inline void virtio_msg_pack_header(VirtIOMSG *msg,
 {
     uint16_t msg_size = virtio_msg_header_size() + payload_size;
 
-    msg->type = type;
+    msg->type = type & VIRTIO_MSG_TYPE_MASK;
     msg->msg_id = msg_id;
     msg->dev_num = cpu_to_le16(dev_num);
     msg->token = cpu_to_le16(token);
@@ -421,11 +493,12 @@ static inline void virtio_msg_pack_get_device_info_resp(VirtIOMSG *msg,
                                                    uint16_t token,
                                                    uint32_t device_id,
                                                    uint32_t vendor_id,
-                                                   uint32_t num_feature_bits,
+                                                   const uint8_t *device_uuid,
+                                                   uint32_t num_feature_blocks,
                                                    uint32_t config_size,
-                                                   uint32_t max_vqs,
-                                                   uint16_t admin_vq_idx,
-                                                   uint16_t admin_vq_count)
+                                                   uint32_t max_virtqueues,
+                                                   uint32_t admin_vq_start,
+                                                   uint32_t admin_vq_count)
 {
     virtio_msg_pack_header(msg, VIRTIO_MSG_DEVICE_INFO,
                            VIRTIO_MSG_TYPE_RESPONSE, dev_num, token,
@@ -433,11 +506,14 @@ static inline void virtio_msg_pack_get_device_info_resp(VirtIOMSG *msg,
 
     msg->get_device_info_resp.device_id = cpu_to_le32(device_id);
     msg->get_device_info_resp.vendor_id = cpu_to_le32(vendor_id);
-    msg->get_device_info_resp.num_feature_bits = cpu_to_le32(num_feature_bits);
+    memcpy(msg->get_device_info_resp.device_uuid, device_uuid,
+           sizeof msg->get_device_info_resp.device_uuid);
+    msg->get_device_info_resp.num_feature_blocks =
+        cpu_to_le32(num_feature_blocks);
     msg->get_device_info_resp.config_size = cpu_to_le32(config_size);
-    msg->get_device_info_resp.max_vqs = cpu_to_le32(max_vqs);
-    msg->get_device_info_resp.admin_vq_idx = cpu_to_le16(admin_vq_idx);
-    msg->get_device_info_resp.admin_vq_count = cpu_to_le16(admin_vq_count);
+    msg->get_device_info_resp.max_virtqueues = cpu_to_le32(max_virtqueues);
+    msg->get_device_info_resp.admin_vq_start = cpu_to_le32(admin_vq_start);
+    msg->get_device_info_resp.admin_vq_count = cpu_to_le32(admin_vq_count);
 }
 
 static inline void virtio_msg_pack_get_features(VirtIOMSG *msg,
@@ -449,8 +525,8 @@ static inline void virtio_msg_pack_get_features(VirtIOMSG *msg,
     virtio_msg_pack_header(msg, VIRTIO_MSG_GET_FEATURES, 0, dev_num, token,
                            sizeof msg->get_features);
 
-    msg->get_features.index = cpu_to_le32(index);
-    msg->get_features.num = cpu_to_le32(num);
+    msg->get_features.block_index = cpu_to_le32(index);
+    msg->get_features.num_blocks = cpu_to_le32(num);
 }
 
 static inline void virtio_msg_pack_get_features_resp(VirtIOMSG *msg,
@@ -466,13 +542,13 @@ static inline void virtio_msg_pack_get_features_resp(VirtIOMSG *msg,
                            VIRTIO_MSG_TYPE_RESPONSE, dev_num, token,
                            sizeof msg->get_features_resp + num * sizeof(*f));
 
-    msg->get_features_resp.index = cpu_to_le32(index);
-    msg->get_features_resp.num = cpu_to_le32(num);
+    msg->get_features_resp.block_index = cpu_to_le32(index);
+    msg->get_features_resp.num_blocks = cpu_to_le32(num);
 
     assert(num <= VIRTIO_MSG_MAX_FEATURE_NUM);
 
     for (i = 0; i < num && i < VIRTIO_MSG_MAX_FEATURE_NUM; i++) {
-        msg->get_features_resp.b32[i] = cpu_to_le32(f[i]);
+        msg->get_features_resp.features[i] = cpu_to_le32(f[i]);
     }
 }
 
@@ -488,13 +564,13 @@ static inline void virtio_msg_pack_set_features(VirtIOMSG *msg,
     virtio_msg_pack_header(msg, VIRTIO_MSG_SET_FEATURES, 0, dev_num, token,
                            sizeof msg->set_features + num * sizeof(*f));
 
-    msg->set_features.index = cpu_to_le32(index);
-    msg->set_features.num = cpu_to_le32(num);
+    msg->set_features.block_index = cpu_to_le32(index);
+    msg->set_features.num_blocks = cpu_to_le32(num);
 
     assert(num <= VIRTIO_MSG_MAX_FEATURE_NUM);
 
     for (i = 0; i < num && i < VIRTIO_MSG_MAX_FEATURE_NUM; i++) {
-        msg->set_features.b32[i] = cpu_to_le32(f[i]);
+        msg->set_features.features[i] = cpu_to_le32(f[i]);
     }
 }
 
@@ -641,6 +717,7 @@ static inline void virtio_msg_pack_get_vqueue_resp(VirtIOMSG *msg,
                                                    uint32_t index,
                                                    uint32_t max_size,
                                                    uint32_t size,
+                                                   uint32_t flags,
                                                    uint64_t descriptor_addr,
                                                    uint64_t driver_addr,
                                                    uint64_t device_addr)
@@ -652,6 +729,7 @@ static inline void virtio_msg_pack_get_vqueue_resp(VirtIOMSG *msg,
     msg->get_vqueue_resp.index = cpu_to_le32(index);
     msg->get_vqueue_resp.max_size = cpu_to_le32(max_size);
     msg->get_vqueue_resp.size = cpu_to_le32(size);
+    msg->get_vqueue_resp.flags = cpu_to_le32(flags);
     msg->get_vqueue_resp.descriptor_addr = cpu_to_le64(descriptor_addr);
     msg->get_vqueue_resp.driver_addr = cpu_to_le64(driver_addr);
     msg->get_vqueue_resp.device_addr = cpu_to_le64(device_addr);
@@ -679,34 +757,36 @@ static inline void virtio_msg_pack_reset_vqueue_resp(VirtIOMSG *msg,
 static inline void virtio_msg_pack_get_shm(VirtIOMSG *msg,
                                            uint16_t dev_num,
                                            uint16_t token,
-                                           uint32_t index)
+                                           uint32_t shmid)
 {
     virtio_msg_pack_header(msg, VIRTIO_MSG_GET_SHM, 0, dev_num, token,
                            sizeof msg->get_shm);
 
-    msg->get_shm.index = cpu_to_le32(index);
+    msg->get_shm.shmid = cpu_to_le32(shmid);
 }
 
 static inline void virtio_msg_pack_get_shm_resp(VirtIOMSG *msg,
                                                 uint16_t dev_num,
                                                 uint16_t token,
-                                                uint32_t index,
-                                                uint32_t length,
-                                                uint32_t address)
+                                                uint32_t shmid,
+                                                uint64_t length,
+                                                uint64_t address)
 {
     virtio_msg_pack_header(msg, VIRTIO_MSG_GET_SHM,
                            VIRTIO_MSG_TYPE_RESPONSE, dev_num, token,
                            sizeof msg->get_shm_resp);
 
-    msg->get_shm_resp.index = cpu_to_le32(index);
-    msg->get_shm_resp.length = cpu_to_le32(length);
-    msg->get_shm_resp.address = cpu_to_le32(address);
+    msg->get_shm_resp.shmid = cpu_to_le32(shmid);
+    msg->get_shm_resp.reserved = 0;
+    msg->get_shm_resp.length = cpu_to_le64(length);
+    msg->get_shm_resp.address = cpu_to_le64(address);
 }
 
 static inline void virtio_msg_pack_set_vqueue(VirtIOMSG *msg,
                                               uint16_t dev_num,
                                               uint16_t token,
                                               uint32_t index,
+                                              uint32_t flags,
                                               uint32_t size,
                                               uint64_t descriptor_addr,
                                               uint64_t driver_addr,
@@ -716,8 +796,9 @@ static inline void virtio_msg_pack_set_vqueue(VirtIOMSG *msg,
                            sizeof msg->set_vqueue);
 
     msg->set_vqueue.index = cpu_to_le32(index);
-    msg->set_vqueue.unused = 0;
+    msg->set_vqueue.flags = cpu_to_le32(flags);
     msg->set_vqueue.size = cpu_to_le32(size);
+    msg->set_vqueue.reserved = 0;
     msg->set_vqueue.descriptor_addr = cpu_to_le64(descriptor_addr);
     msg->set_vqueue.driver_addr = cpu_to_le64(driver_addr);
     msg->set_vqueue.device_addr = cpu_to_le64(device_addr);
@@ -771,10 +852,8 @@ static inline void virtio_msg_pack_event_config(VirtIOMSG *msg,
                                                 uint32_t size,
                                                 uint8_t *value)
 {
-    unsigned int max_size;
+    unsigned int max_size = VIRTIO_MSG_MAX_EVENT_CONFIG_BYTES;
 
-    max_size = VIRTIO_MSG_MAX_SIZE;
-    max_size -= offsetof(VirtIOMSG, event_config.config_value);
     assert(size <= max_size);
 
     virtio_msg_pack_header(msg, VIRTIO_MSG_EVENT_CONFIG, 0, dev_num, 0,
@@ -804,8 +883,9 @@ static inline void virtio_msg_pack_bus_get_devices(VirtIOMSG *msg,
 
 static inline void virtio_msg_pack_bus_get_devices_resp(VirtIOMSG *msg,
                                                         uint16_t offset,
-                                                        uint16_t num,
                                                         uint16_t next_offset,
+                                                        uint16_t num,
+                                                        uint16_t token,
                                                         uint8_t *data)
 {
     size_t byte_len = (num + 7) / 8;
@@ -813,11 +893,12 @@ static inline void virtio_msg_pack_bus_get_devices_resp(VirtIOMSG *msg,
     virtio_msg_pack_header(msg,
                            VIRTIO_MSG_BUS_GET_DEVICES,
                            VIRTIO_MSG_TYPE_BUS | VIRTIO_MSG_TYPE_RESPONSE,
-                           0, 0, sizeof msg->bus_get_devices_resp + byte_len);
+                           0, token,
+                           sizeof msg->bus_get_devices_resp + byte_len);
 
     msg->bus_get_devices_resp.offset = cpu_to_le16(offset);
-    msg->bus_get_devices_resp.num = cpu_to_le16(num);
     msg->bus_get_devices_resp.next_offset = cpu_to_le16(next_offset);
+    msg->bus_get_devices_resp.num = cpu_to_le16(num);
 
     if (byte_len > 0) {
         memcpy(msg->bus_get_devices_resp.data, data, byte_len);
@@ -825,12 +906,13 @@ static inline void virtio_msg_pack_bus_get_devices_resp(VirtIOMSG *msg,
 }
 
 static inline void virtio_msg_pack_bus_ping_resp(VirtIOMSG *msg,
+                                                 uint16_t token,
                                                  uint32_t data)
 {
     virtio_msg_pack_header(msg,
                            VIRTIO_MSG_BUS_PING,
                            VIRTIO_MSG_TYPE_BUS | VIRTIO_MSG_TYPE_RESPONSE,
-                           0, 0, sizeof msg->bus_ping_resp);
+                           0, token, sizeof msg->bus_ping_resp);
 
     msg->bus_ping_resp.data = cpu_to_le32(data);
 }
@@ -861,12 +943,15 @@ static inline const char *virtio_msg_id_to_str(uint8_t type, uint8_t msg_id)
         VIRTIO_MSG_TYPE2STR(GET_VQUEUE),
         VIRTIO_MSG_TYPE2STR(SET_VQUEUE),
         VIRTIO_MSG_TYPE2STR(RESET_VQUEUE),
+        VIRTIO_MSG_TYPE2STR(GET_SHM),
         VIRTIO_MSG_TYPE2STR(EVENT_CONFIG),
         VIRTIO_MSG_TYPE2STR(EVENT_AVAIL),
         VIRTIO_MSG_TYPE2STR(EVENT_USED),
     };
     static const char *bus_type2str[VIRTIO_MSG_MAX + 1] = {
         VIRTIO_MSG_TYPE2STR(BUS_GET_DEVICES),
+        VIRTIO_MSG_TYPE2STR(BUS_PING),
+        VIRTIO_MSG_TYPE2STR(BUS_EVENT_DEVICE),
     };
 
     if (type & VIRTIO_MSG_TYPE_BUS) {
@@ -937,9 +1022,10 @@ static inline void virtio_msg_print(VirtIOMSG *msg)
         virtio_msg_print_status(msg->set_device_status.status);
         break;
     case VIRTIO_MSG_SET_VQUEUE:
-        printf("set-vqueue: index=%d size=%d desc-addr=%lx "
+        printf("set-vqueue: index=%d flags=%x size=%d desc-addr=%lx "
                "driver-addr=%lx device-addr=%lx\n",
-               msg->set_vqueue.index, msg->set_vqueue.size,
+               msg->set_vqueue.index, msg->set_vqueue.flags,
+               msg->set_vqueue.size,
                msg->set_vqueue.descriptor_addr,
                msg->set_vqueue.driver_addr,
                msg->set_vqueue.device_addr);
-- 
2.53.0

